features/lorawan/lorastack/phy/LoRaPHYAS923.cpp

```diff
@@ -0,0 +1,1338 @@
+    getPhy.Attribute = PHY_MIN_TX_DR;
+    getPhy.UplinkDwellTime = adrNext->UplinkDwellTime;
+    phyParam = get_phy_params(&getPhy);
+    minTxDatarate = phyParam.Value;
+
+    // Report back the adr ack counter
+    *adrAckCounter = adrNext->AdrAckCounter;
+
+    // Apply the minimum possible datarate.
+    datarate = MAX( datarate, minTxDatarate );
+
+    if( adrNext->AdrEnabled == true )
+    {
+        if( datarate == minTxDatarate )
+        {
+            *adrAckCounter = 0;
+            adrAckReq = false;
+        }
+        else
+        {
+            if( adrNext->AdrAckCounter >= AS923_ADR_ACK_LIMIT )
+            {
+                adrAckReq = true;
+                txPower = AS923_MAX_TX_POWER;
+            }
+            else
+            {
+                adrAckReq = false;
+            }
+            if( adrNext->AdrAckCounter >= ( AS923_ADR_ACK_LIMIT + AS923_ADR_ACK_DELAY ) )
+            {
+                if( ( adrNext->AdrAckCounter % AS923_ADR_ACK_DELAY ) == 1 )
+                {
+                    // Decrease the datarate
+                    getPhy.Attribute = PHY_NEXT_LOWER_TX_DR;
+                    getPhy.Datarate = datarate;
+                    getPhy.UplinkDwellTime = adrNext->UplinkDwellTime;
+                    phyParam = get_phy_params(&getPhy);
+                    datarate = phyParam.Value;
+
+                    if( datarate == minTxDatarate )
+                    {
+                        // We must set adrAckReq to false as soon as we reach the lowest datarate
+                        adrAckReq = false;
+                        if( adrNext->UpdateChanMask == true )
+                        {
+                            // Re-enable default channels
+                            ChannelsMask[0] |= LC( 1 ) + LC( 2 );
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    *drOut = datarate;
+    *txPowOut = txPower;
+    return adrAckReq;
+}
+
+void LoRaPHYAS923::compute_rx_win_params(int8_t datarate, uint8_t minRxSymbols,
+                                         uint32_t rxError,
+                                         RxConfigParams_t *rxConfigParams)
+{
+    double tSymbol = 0.0;
+
+    // Get the datarate, perform a boundary check
+    rxConfigParams->Datarate = MIN( datarate, AS923_RX_MAX_DATARATE );
+    rxConfigParams->Bandwidth = GetBandwidth( rxConfigParams->Datarate );
+
+    if( rxConfigParams->Datarate == DR_7 )
+    { // FSK
+        tSymbol = compute_symb_timeout_fsk( DataratesAS923[rxConfigParams->Datarate] );
+    }
+    else
+    { // LoRa
+        tSymbol = compute_symb_timeout_lora( DataratesAS923[rxConfigParams->Datarate], BandwidthsAS923[rxConfigParams->Datarate] );
+    }
+
+    get_rx_window_params( tSymbol, minRxSymbols, rxError, RADIO_WAKEUP_TIME, &rxConfigParams->WindowTimeout, &rxConfigParams->WindowOffset );
+}
+
+bool LoRaPHYAS923::rx_config(RxConfigParams_t* rxConfig, int8_t* datarate)
+{
```

Comment:
Again check for rxConfg == NULL?

Reply:
Same as above. Mac layer takes care of it